src/frontend/src/optimizer/plan_node/logical_hop_window.rs

```diff
@@ -317,8 +322,21 @@ impl ColPrunable for LogicalHopWindow {
 
 impl PredicatePushdown for LogicalHopWindow {
     fn predicate_pushdown(&self, predicate: Condition) -> PlanRef {
-        // TODO: hop's predicate pushdown https://github.com/risingwavelabs/risingwave/issues/6606
-        gen_filter_and_pushdown(self, predicate, Condition::true_cond())
+        // Do not push down `window_start`, `window_end` for now, they are columns produced by
+        // HopWindow.
+        let mut window_columns = FixedBitSet::with_capacity(self.schema().len());
+        let window_start_idx = self.input().schema().len();
+        let window_end_idx = self.input().schema().len() + 1;
+        for (i, v) in self.output_indices().iter().enumerate() {
+            if *v == window_start_idx || *v == window_end_idx {
+                window_columns.insert(i);
+            }
+        }
+
+        // Keep predicate on time window (time_window_pred), the rest (pushed_predicate) may be
+        // pushed-down.
+        let (time_window_pred, pushed_predicate) = predicate.split_disjoint(&window_columns);
```

Comment:
I think we should rewrite the `pushed_predicate` with the mapping from output to input, which can be calculated by the `self.output_indices`. Though the `output_indices` here is just to express the selection of the window_start_idx and window_end_idx, we can not assume that here.
(Maybe this tells that the output indices here is not a good abstraction, but in this PR we must do it.)

Reply:
But we do not need `Substitute` here. just using `self.o2i_col_mapping()` to rewrite the predicate is ok.